src/main/java/net/apnic/whowas/types/Parsing.java

```diff
@@ -1,29 +1,62 @@
 package net.apnic.whowas.types;
 
-import org.codehaus.jparsec.Parser;
-import org.codehaus.jparsec.Parsers;
-import org.codehaus.jparsec.Scanners;
-import org.codehaus.jparsec.pattern.CharPredicates;
-import org.codehaus.jparsec.pattern.Patterns;
-
 import java.net.Inet4Address;
 import java.net.Inet6Address;
 import java.net.UnknownHostException;
 import java.util.OptionalInt;
 
+import org.codehaus.jparsec.Parser;
+import org.codehaus.jparsec.Parsers;
+import org.codehaus.jparsec.pattern.CharPredicate;
+import org.codehaus.jparsec.pattern.CharPredicates;
+import org.codehaus.jparsec.pattern.Pattern;
+import org.codehaus.jparsec.pattern.Patterns;
+import org.codehaus.jparsec.Scanners;
+
 public final class Parsing {
+    private static final CharPredicate IPV6_SEC_PREDICATE =
+        CharPredicates.among("0123456789abcdefABCDEF");
+
+    private static final Pattern IPV6_SEC_PATTERN =
+        Patterns.times(1, 4, IPV6_SEC_PREDICATE);
+
+    private static final Pattern IPV6_SEC_COLON_PATTERN =
+        IPV6_SEC_PATTERN.next(Patterns.isChar(':'));
+
+    private static final Pattern IPV6_DCOLON_END_PATTERN =
+        IPV6_SEC_COLON_PATTERN.times(0, 5).next(IPV6_SEC_PATTERN)
+            .or(Patterns.times(0, 4, IPV6_SEC_PREDICATE));
+
+    private static final Pattern IPV6_ALL_SEC_PATTERN =
+        IPV6_SEC_COLON_PATTERN.times(7).next(IPV6_SEC_PATTERN);
+
+    private static final Pattern IPV6_PART_SEC_PATTERN =
+        IPV6_SEC_COLON_PATTERN.times(1, 6).or(Patterns.isChar(':'))
+            .next(Patterns.isChar(':'))
+            .next(IPV6_DCOLON_END_PATTERN);
+
+    private static final Pattern IPV6_FULL_PATTERN =
+        IPV6_PART_SEC_PATTERN.or(IPV6_ALL_SEC_PATTERN);
+
     private Parsing() {}
 
     private static final Parser<IP> V4_ADDRESS =
-            Patterns.many(CharPredicates.among("0123456789.")).toScanner("IPv4 address").source().map(s -> {
+        Patterns.INTEGER
+            .next(Patterns.isChar('.'))
+            .times(3)
+            .next(Patterns.INTEGER).toScanner("IPv44 Address").source()
+            .map(s ->
+            {
                 try {
                     return new IP(Inet4Address.getByName(s));
                 } catch (UnknownHostException e) {
                     throw new RuntimeException(e);
```

Comment:
{suggestion} Maybe a comment somewhere saying that the unknown host exception is only expected if a non-ip address is given to Inet4/6Address? (which the supplied pattern should prevent) Seeing the RuntimeException wrapper here threw me as I would have expected a ParseException (or similar).